display_list/display_list.h

```diff
@@ -138,14 +138,16 @@ namespace flutter {
   V(DrawShadowTransparentOccluder)
 
 #define DL_OP_TO_ENUM_VALUE(name) k##name,
-enum class DisplayListOpType {
+enum class DisplayListOpType : uint8_t {
```

Comment:
The only place this is used, I believe, (at least used in a place where we care about the size) already has a size specifier on it to keep it to 8 bits.

Reply:
Yep, it's marked as 8 bits in the DlOp.  C++ isn't complaining about throwing away the bits.  This is safer since we won't be throwing away bits anymore.

Reply:
Without this change sizeof(DisplayListOpType) is 4.  I guess because it's a class enum though it may be able to determine that throwing away those bits isn't an issue.

Reply:
Right, we don't have an assert that we are using less than 256 values. I guess I don't have an issue, but more that I'm wondering why we are worried about this all of a sudden.

Reply:
You mean why do I bother to read the engine code critically?  =)

Reply:
Hmm, looks like this broke some tests.  I'm guessing somewhere in the test code we were using sizeof(DisplayListOpType) instead of sizeof(DlOp)?  I can't see why, seems like a bug in testing code, but it isn't apparent where.  If it isn't obvious to you we can drop this.

Reply:
> You mean why do I bother to read the engine code critically? =)

No, that's a good thing, I'm referring more to priority.

Reply:
> Hmm, looks like this broke some tests. I'm guessing somewhere in the test code we were using sizeof(DisplayListOpType) instead of sizeof(DlOp)? I can't see why, seems like a bug in testing code, but it isn't apparent where. If it isn't obvious to you we can drop this.

It looks like it's mainly Windows which I've always had trouble with. Given that this change causes a huge increase in the storage sizes (as seen in the output of the results of the SingleOp tests), I wonder if the problem is how it applies the bitfield specifiers on this line: https://github.com/flutter/engine/blob/f9c497f178d3a8a8b77731c8a15ea46b558b0081/display_list/dl_op_records.h#L102

Perhaps it will bitfield-pack an enum into 8 bits, but when the type inherits from `uint8_t` then it decides to break the bitfields apart so before it had 8 bits of type and 24 bits of size == 32 bits, but now it has "uh, oh, that's a uint8_t, I can't pack that so I'll give it its own byte" followed by "and now they want a size field packed into 24 bits? Well that's an entirely new thing so I'll align this part to the next word boundary". I could run some experiments with godbolt and our Windows compilers to see if they suddenly refuse to pack the fields. I've seen similar cases in the past where it makes pessimal assumptions about packing that seem totally unnecessary but it ends up wasting a lot of bytes in the structure packing.

The ordering of the fields in the ops were designed to eliminate as much wasted packing space as they could so suddenly having the combined type/size bitfields expand from 4 to 8 bytes means that the ordering of the fields in the op records after that might break. Typically I'll try to have the first thing in an op record by a 32-bit quantity so that if it is followed by a 64-bit (or longer) aligned field then we'll use those 32-bits rather than skip them. If the end of the type/size bits is now 64-bit aligned then all of the alignments after that will be out of phase...

Reply:
One could also ask if all of this size testing is really necessary, or if it is a premature micro-optimization, but this tiny change points out that what looks like "good C++ code" ends up causing Windows to waste a bunch of bits in the structs...